### DISCUSSION TITLE
Question about Alias analysis

### DISCUSSION THREAD

**Discussion by c8ef** [dotnet/runtime#96434] [Q&A] [answered]
Hi, I noticed that according to [this document](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/ryujit-overview.md), Ryujit has implemented the dead store elimination optimization. However, it appears that it does not implement Alias analysis like LLVM does in its compiler backend. I'm curious about the difference between having or not having alias analysis for DSE.

**Comment by SingleAccretion** [comment, marked as answer]
Indeed, RyuJit doesn't implement general LLVM-style alias analysis. It does have a scoped version as [part of the value numbering](https://github.com/dotnet/runtime/blob/23e5d5410ed3e8bc9ab61b3685e1d3a06bdba701/src/coreclr/jit/valuenum.h#L22-L146).

It does mean we don't dead-store writes to non-local memory. In a sense, the more general analysis is important in LLVM because it represents local memory modification via stores to `alloca`s, while in RyuJit both enregisterable values and things like structs are `LCL_VAR`s. And, of course, LLVM has a very complex pointer aliasing model that is intended to take advantage of the optimization opportunities presented by liberal UB rules in C/C++, something that does not exist in IL.

**Comment by c8ef** [comment]
That's cool, thanks for your explanation!
